logstash-core/src/main/java/org/logstash/ackedqueue/FsUtil.java

```diff
@@ -0,0 +1,47 @@
+package org.logstash.ackedqueue;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * File System Utility Methods.
+ */
+public final class FsUtil {
+
+    private FsUtil() {
+    }
+
+    /**
+     * Checks if the request number of bytes of free disk space are available under the given
+     * path.
+     * @param path Directory to check
+     * @param size Bytes of free space requested
+     * @return True iff the
+     * @throws IOException on failure to determine free space for given path's partition
+     */
+    public static boolean hasFreeSpace(final String path, final long size) throws IOException {
+        final File[] partitions = File.listRoots();
+        File location = new File(path).getCanonicalFile();
+        boolean found = false;
+        while (!found) {
+            for (final File partition : partitions) {
+                if (partition.equals(location)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                location = location.getParentFile();
+                if (location == null) {
+                    throw new IllegalStateException(
```

Comment:
@colinsurprenant I don't have a very strong opinion on this one, but my worry is essentially this situation:

* User uses some strange path (maybe some `NFS` style thing) that doesn't show up with a partition that the JVM understands and we throw a checked Exception that may get caught by upstream logic (or worse yet re-thrown in some opaque way) => Likely it will be easier to understand this in a support case if we have it be an unchecked one that just bubbles up directly (also philosophically it makes more sense to throw unchecked for situations that are deemed impossible)

... again no strong opinion on this though :) the below point seems more important :)